cluster_experiments/perturbator.py

```diff
@@ -90,6 +91,217 @@ def perturbate(
+                "Simulated effect needs to be bigger than -100%, got "
+                f"{average_effect*100:.1f}%"
+            )
+
+        treatment_zeros = (
+            (df[self.treatment_col] != self.treatment) | (df[self.target_col] == 0)
+        ).mean()
+        if 1.0 == treatment_zeros:
+            raise ValueError(
+                f"All treatment samples have {self.target_col} = 0, relative effect "
+                f"{average_effect} will have no effect"
+            )
+
+    def _apply_multiplicative_effect(
+        self, df: pd.DataFrame, effect: Union[float, np.ndarray]
+    ) -> pd.DataFrame:
+        df.loc[df[self.treatment_col] == self.treatment, self.target_col] *= 1 + effect
+        return df
+
+
+class StochasticRelativePositivePerturbator(
+    StochasticPerturbator, RelativePositivePerturbator
+):
+    """
+    A Perturbator for continuous, positively-defined targets that applies a simulated
+    stochastic effect multiplicatively for the treated samples, ie. proportional to the
+    target value for each sample. The number of samples with 0 as target remains unchanged.
+
+    The stochastic effect is sampled from a beta distribution with parameters mean and
+    variance. If variance is not provided, the variance is abs(mean).
+
+    The common beta parameters are derived from the mean and scale as follows:
+
+    ```
+    a <- mu / (scale * scale)
+    b <- (1-mu) / (scale * scale)
+    effect ~ beta(a, b)
+    ```
+    source: https://stackoverflow.com/a/51143208
+
+    Example: a mean = 0.2 and variance = 0.1, give a = 20 and b = 80
+    Plot: https://www.wolframalpha.com/input?i=plot+distribution+of+beta%2820%2C+80%29
+
+    Arguments:
+        average_effect (Optional[float], optional): the average effect of the treatment. Defaults to None.
+        target_col (str, optional): name of the target_col to use as the outcome. Defaults to "target".
+        treatment_col (str, optional): the name of the column that contains the treatment. Defaults to "treatment".
+        treatment (str, optional): name of the treatment to use as the treated group. Defaults to "B".
+        scale (Optional[float], optional): the scale of the effect distribution. Defaults to None.
+    """
+
+    def perturbate(
+        self, df: pd.DataFrame, average_effect: Optional[float] = None
+    ) -> pd.DataFrame:
+        """
+        Usage:
+        ```python
+        from cluster_experiments.perturbator import StochasticRelativePositivePerturbator
+        import pandas as pd
+        df = pd.DataFrame({"target": [1, 2, 3], "treatment": ["A", "B", "A"]})
+        perturbator = StochasticRelativePositivePerturbator()
+        # Increase target metric by 20%
+        perturbator.perturbate(df, average_effect=0.2)
+        # returns pd.DataFrame({"target": [1, 3, 3], "treatment": ["A", "B", "A"]})
+        ```
+        """
+        df = df.copy().reset_index(drop=True)
+        average_effect = self.get_average_effect(average_effect)
```

Comment:
Perhaps out of the scope of this PR, but since you're working on the perturbators, wdyt about renaming get_average_effect to _get_average_effect?

Reply:
(for all perturbators)

Reply:
Thanks for the comment, since it's invoked by other classes, it can't be a private method. Therefore, it's better like this.